frontend/src/contexts/UndoRedoContext.tsx

```
@@ -0,0 +1,311 @@
+
+  const debounceMap = useRef<Map<string, number>>(new Map());
+
+  const persistActions = useCallback(() => {
+    try {
+      const data: { undoStack: SerializableActionEntry[]; redoStack: SerializableActionEntry[] } = {
+        undoStack: stateRef.current.undoStack.map(serializeAction),
+        redoStack: stateRef.current.redoStack.map(serializeAction),
+      };
+      localStorage.setItem(persistKey, JSON.stringify(data));
+    } catch {
+    }
```

ESLint

[error] Empty block statement.

(no-empty)